### DISCUSSION TITLE
Exif Class usage

### DISCUSSION THREAD

**Discussion by john-giles-ff** [dlemstra/Magick.NET#1708] [Help]
Hello,

I am new to the Magick.net library. It is quite an impressive piece of work.
I am working on a project where the new library has been updated to Magick.Net-Q16-AnyCPU 13.6.0

I am having trouble converting the code to work with the new library. It is mainly to do with the new Exif class and being able to set a new profile value.

The is way it was working before

            var profile = new ExifProfile();
            foreach (var exifEntry in image.ExifData)
            {
                try
                {
                    if (_enum.IsDefined(typeof(ExifTag), exifEntry.Id))
                    {
                        var tag = (ExifTag)exifEntry.Id;

                        object value = exifEntry.Value;

                        if ((tag == ExifTag.XResolution || tag == ExifTag.YResolution || tag == ExifTag.ExposureTime || tag == ExifTag.DigitalZoomRatio || tag == ExifTag.MaxApertureValue || tag == ExifTag.FNumber || tag == ExifTag.FocalLength || tag == ExifTag.XResolution || tag == ExifTag.YResolution || tag == ExifTag.ApertureValue))
                        {
                            if (exifEntry.Value.ToString().Contains("/"))
                            {
                                var numbers = exifEntry.Value.ToString().Split('/').Select(a => uint.Parse(a)).ToArray();
                                value = new Rational(numbers[0], numbers[1], true);
                            }
                            else ...
                           
with profile.SetValue(tag, value);
at the end.

Now I have

            var profile = new ExifProfile();
            foreach (var exifEntry in image.ExifData)
            {
                try
                {
                    if (_enum.IsDefined(typeof(ExifTagValue), exifEntry.Id))
                    {
                        var tag = (ExifTagValue)exifEntry.Id;
                        object value = exifEntry.Value;

                        var magickTag = (ExifTag<ExifTagValue>)(typeof(ExifTag).GetProperty(tag.ToString()).GetValue(null, null));

                        if ((tag == ExifTagValue.XResolution || tag == ExifTagValue.YResolution || tag == ExifTagValue.ExposureTime || tag == ExifTagValue.DigitalZoomRatio || tag == ExifTagValue.MaxApertureValue || tag == ExifTagValue.FNumber || tag == ExifTagValue.FocalLength || tag == ExifTagValue.XResolution || tag == ExifTagValue.YResolution || tag == ExifTagValue.ApertureValue))
                        {
                            if (exifEntry.Value.ToString().Contains("/"))
                            {
                                var numbers = exifEntry.Value.ToString().Split('/').Select(a => uint.Parse(a)).ToArray();
                                value = new Rational(numbers[0], numbers[1], true);
                            }
                            else ...

The magickTag var is not right but I cannot see how to satisfy the command

profile.SetValue<ExifTagValue>(magickTag, (ExifTagValue)value);

It may be my misunderstanding of how to use the new Exif class. I have looked through all the readme files.
Please could you point me in the right direction?

Kind regards

John

**Comment by dlemstra** [comment]
The second argument (`value`) should have the correct type that the specified for that specified `ExifTag` so depending on the type you need to typecast it. I suspect that in your case it should be `(Rational)value)`.

**Reply by john-giles-ff** [reply]
Thank you for your response. In most cases it is Rational, however sometime it is number, Signed Rational or even byte[]
I have got a solution for now. What I have done is create a method call to determine the correct ExifTag form the passed ExifTagValue so I know what to cast the object. 
There may be a better way of determining the value type and ExifTag from the ExifTagValue but I cannot think of on right now.

I have placed the code below for your interest.

Thank you again for your prompt response.

So instead of of placing 
profile.SetValue(magickTag, (ExifTagValue)value);

at the end of the if else statements that parse the ExifTagValue tag into an object I call

ProfileSetValue(profile, tag, value);

Method below.

   private void ProfileSetValue(ExifProfile profile, ExifTagValue tag, object value)
   {
       switch(tag)
       {
           case ExifTagValue.XResolution:
               profile.SetValue(ExifTag.XResolution, (Rational)value);
               break; ;

           case ExifTagValue.YResolution:
               profile.SetValue(ExifTag.YResolution, (Rational)value);
               break; ;

           case ExifTagValue.ExposureTime:
               profile.SetValue(ExifTag.ExposureTime, (Rational)value);
               break; ;

           case ExifTagValue.DigitalZoomRatio:
               profile.SetValue(ExifTag.DigitalZoomRatio, (Rational)value);
               break; ;

           case ExifTagValue.MaxApertureValue:
               profile.SetValue(ExifTag.MaxApertureValue, (Rational)value);
               break; ;

           case ExifTagValue.FNumber:
               profile.SetValue(ExifTag.XResolution, (Rational)value);
               break; ;

           case ExifTagValue.FocalLength:
               profile.SetValue(ExifTag.FocalLength, (Rational)value);
               break; ;

           case ExifTagValue.ApertureValue:
               profile.SetValue(ExifTag.ApertureValue, (Rational)value);
               break; ;

           case ExifTagValue.ISOSpeedRatings:
               profile.SetValue(ExifTag.ISOSpeedRatings, (ushort[])value);
               break; ;

           case ExifTagValue.ShutterSpeedValue:
               profile.SetValue(ExifTag.ShutterSpeedValue, (SignedRational)value);
               break; ;

           case ExifTagValue.BrightnessValue:
               profile.SetValue(ExifTag.BrightnessValue, (SignedRational)value);
               break; ;

           case ExifTagValue.ExposureBiasValue:
               profile.SetValue(ExifTag.ExposureBiasValue, (SignedRational)value);
               break; ;

           case ExifTagValue.UserComment:
               profile.SetValue(ExifTag.UserComment, (byte[])value);
               break; ;

           case ExifTagValue.ExifVersion:
               profile.SetValue(ExifTag.ExifVersion, (byte[])value);
               break; ;

           case ExifTagValue.GPSLatitude:
               profile.SetValue(ExifTag.GPSLatitude, (Rational[])value);
               break; ;

           case ExifTagValue.GPSLongitude:
               profile.SetValue(ExifTag.GPSLongitude, (Rational[])value);
               break; ;

           case ExifTagValue.ResolutionUnit:
               profile.SetValue(ExifTag.ResolutionUnit, (ushort)value);
               break; ;

           case ExifTagValue.ImageWidth:
               profile.SetValue(ExifTag.ImageWidth, (Number)value);
               break; ;

           case ExifTagValue.ImageLength:
               profile.SetValue(ExifTag.ImageLength, (Number)value);
               break; ;
       }

   }

**Reply by dlemstra** [reply]
Thanks for posting here how you solved it for future readers.
